src/backend/InvenTree/report/models.py

```diff
@@ -501,6 +518,47 @@ def print(self, items: list, request=None, output=None, **kwargs) -> DataOutput:
                 # Update the progress of the report generation
                 output.progress += 1
                 output.save()
+            else:
+                for instance in items:
+                    context = self.get_context(instance, request)
+
+                    if report_name is None:
+                        report_name = self.generate_filename(context)
+
+                    # Render the report output
+                    try:
+                        if debug_mode:
+                            report = self.render_as_string(instance, request)
+                        else:
+                            report = self.render(instance, request)
+                    except TemplateDoesNotExist as e:
+                        t_name = str(e) or self.template
+                        raise ValidationError(f'Template file {t_name} does not exist')
+
+                    outputs.append(report)
+
+                    # Attach the generated report to the model instance (if required)
+                    if self.attach_to_model and not debug_mode:
+                        instance.create_attachment(
+                            attachment=ContentFile(report, report_name),
+                            comment=_(f'Report generated from template {self.name}'),
+                            upload_user=request.user
+                            if request and request.user.is_authenticated
+                            else None,
+                        )
+
+                    # Provide generated report to any interested plugins
+                    for plugin in report_plugins:
+                        try:
+                            plugin.report_callback(self, instance, report, request)
+                        except Exception:
+                            InvenTree.exceptions.log_error(
+                                f'plugins.{plugin.slug}.report_callback'
+                            )
+
+                    # Update the progress of the report generation
+                    output.progress += 1
+                    output.save()
```

Comment:
There is a lot duplication between the 2 sections; can you try reducing that? Every change in rendering logic going forward will have to be done twice.